FrontendTool: fix escaping of filenames in Makefile

compnerd opened the pull request:
The dependency file that is being generated should not escape : and # in
the filename.  This makes the behaviour of the filename escaping similar
to clang and GCC and fixes incorrect quoting of paths on Windows.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd commented:
CC: @jrose-apple

compnerd commented:
@swift-ci please test

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -122,18 +122,26 @@ static bool emitMakeDependenciesIfNeeded(DiagnosticEngine &diags,
+
+      case ' ':
+      case '\t':
+        for (unsigned J = I; J && raw[J - 1] == '\\'; --J)
```
What if `I` is 0? Also, what is this defending against?

compnerd replied:
`J &&` would be false there, and not enter the loop.

compnerd replied:
Err, yes.  Added a comment explaining whats going on there.

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -122,18 +122,26 @@ static bool emitMakeDependenciesIfNeeded(DiagnosticEngine &diags,
+        pathBuf.push_back('$');
+        break;
+      }
+      patBuf.push_back(raw[I]);
```
Typo: `pathBuf`.

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -122,18 +122,26 @@ static bool emitMakeDependenciesIfNeeded(DiagnosticEngine &diags,
-    static const char badChars[] = " $#:\n";
-    size_t prev = 0;
-    for (auto index = raw.find_first_of(badChars); index != StringRef::npos;
-         index = raw.find_first_of(badChars, index+1)) {
```
The purpose of this was to make the loop faster by doing bulk copies of most characters. Is that not worth it?

compnerd replied:
I don't think that it will matter too much in use.  clang (and gcc) both do this.  Plus, if we try to be clever, we need to backtrack in the case that there is a space in the name, and I expect that isn't that unreasonable.  If you think that added complexity is worth it, I can try to revert to that.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

jrose-apple commented:
Code looks good, please add a test. (Or extract it out to a non-static function and use gtest, if that's easier for setting up bad file names.)

compnerd commented:
@jrose-apple - hmm, extract the `escape` lambda you mean?  That would allow us to test that with a bunch of "bad" filenames.

jrose-apple commented:
I meant the whole function that takes a DependencyTracker, but just the `escape` bit would be reasonable too. (`escapeForMakefile` or something.)

compnerd committed:
FrontendTool: fix escaping of filenames in Makefile

The dependency file that is being generated should not escape : and # in
the filename.  This makes the behaviour of the filename escaping similar
to clang and GCC and fixes incorrect quoting of paths on Windows.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -99,6 +99,44 @@ static std::string displayName(StringRef MainExecutablePath) {
+namespace swift {
+namespace frontend {
+namespace utils {
+StringRef quoteForMake(StringRef raw, llvm::SmallVectorImpl<char> &buffer) {
```
Nitpick: can you do this with `::` instead of nested namespaces?

jrose-apple replied:
Also, maybe `escapeForMake` to go with `llvm::yaml::escape`, especially since we're not using quotes.

compnerd replied:
Sure, both are fine by me; will do.

compnerd commented:
@swift-ci please test and merge

compnerd commented:
@swift-ci please test and merge
